pubsub/google/cloud/pubsub_v1/publisher/future.py

```diff
@@ -0,0 +1,150 @@
+        This still returns True in failure cases; checking `result` or
+        `exception` is the canonical way to assess success or failure.
+        """
+        return self._batch_info.status in ('success', 'error')
+
+    def result(self, timeout=None):
+        """Return the message ID, or raise an exception.
+
+        This blocks until the message has successfully been published, and
+        returns the message ID.
+
+        Args:
+            timeout (int|float): The number of seconds before this call
+                times out and raises TimeoutError.
+
+        Raises:
+            :class:~`pubsub_v1.TimeoutError`: If the request times out.
+            :class:~`Exception`: For undefined exceptions in the underlying
+                call execution.
+        """
+        # Attempt to get the exception if there is one.
+        # If there is not one, then we know everything worked, and we can
+        # return an appropriate value.
+        err = self.exception(timeout=timeout)
+        if err is None:
+            return self._batch_info.message_ids[hash(self)]
+        raise err
+
+    def exception(self, timeout=None, _wait=1):
+        """Return the exception raised by the call, if any.
+
+        This blocks until the message has successfully been published, and
+        returns the exception. If the call succeeded, return None.
+
+        Args:
+            timeout (int|float): The number of seconds before this call
+                times out and raises TimeoutError.
+
+        Raises:
+            :exc:`TimeoutError`: If the request times out.
+
+        Returns:
+            :class:`Exception`: The exception raised by the call, if any.
+        """
+        # If the batch completed successfully, this should return None.
+        if self._batch_info.status == 'success':
+            return None
+
+        # If this batch had an error, this should return it.
+        if self._batch_info.status == 'error':
+            return self._batch_info.error
+
+        # If the timeout has been exceeded, raise TimeoutError.
+        if timeout and timeout < 0:
+            raise TimeoutError('Timed out waiting for an exception.')
+
+        # Wait a little while and try again.
+        time.sleep(_wait)
+        return self.exception(
+            timeout=timeout - _wait,
+            _wait=min(_wait * 2, 60),
+        )
+
+    def add_done_callback(self, fn):
+        """Attach the provided callable to the future.
+
+        The provided function is called, with this future as its only argument,
+        when the future finishes running.
+        """
+        if self.done():
+            fn(self)
+        self._callbacks.append(fn)
+
+    def _trigger(self):
+        """Trigger all callbacks registered to this Future.
+
+        This method is called internally by the batch once the batch
+        completes.
+
+        Args:
+            message_id (str): The message ID, as a string.
+        """
+        for callback in self._callbacks:
```

Comment:
@lukesneeringer There is a 2.7 [backport][1]

[1]: https://pypi.python.org/pypi/futures/3.1.1